tagging_accuracy/tagging_accuracy_report.py

```diff
@@ -108,7 +115,54 @@ def match_tags(self):
             f.write(json.dumps(no_tags_list))
         with open('partially_correct.json', 'w') as pc_json:
             pc_json.write(json.dumps(partially_correct_list, indent=4, sort_keys=True))
-        return no_tags, man_tags_not_in_result, man_tags_not_collected, total_packages, correctly_tagged, partially_correct
+        return (no_tags, man_tags_not_in_result, man_tags_not_collected, total_packages,
+                correctly_tagged, partially_correct)
+
+    def check_tag_matches(self, num_tags_in_result=4):
+        """Check the common tags between two packages and aggregate metric.
+
+        This metric is of importance from the point of view of the PGM as for us to give better
+        alternate and companion package recommendations we need that our packages be grouped
+        under common topics.
+        """
+        occurences = {}
+        # We need to compare each package to every other package in order to make sure that we
+        # cover all bases. At the same time we need to make sure that we don't consider any
+        # combo twice.
+        automated_tags_dict = self.automated_tags_dict
+        for package_name in list(automated_tags_dict):
+            current_package_tags = automated_tags_dict.pop(package_name)
+            # copy the list so as to not alter it
+            automated_tags_copy_list = automated_tags_dict.values()
+            res = list(map(lambda this_package_tags: len(this_package_tags - current_package_tags),
+                           automated_tags_copy_list))
+            for uncommon_count in res:
+                # if we find common tags between two packages
+                if uncommon_count < 4:
+                    common_count = num_tags_in_result - uncommon_count
+                    occurences[common_count] = \
+                        occurences.get(common_count, 0) + 1
+        for common_count in occurences:
+            print("No. of package pairs with {} common tags: {}".format(common_count,
+                                                                        occurences[common_count]))
+
+    def _calculate_euclidean_distance(self, partially_correct_list):
+        """Calculates the euclidean distance based on the position at which the tag occurs
+        from the top in _sorted."""
+        total_sum = 0
+        for package_name in partially_correct_list:
+            # check which tags are not a part of the result
+            tags_missing = (set(self.package_topic_json[package_name]) -
+                            set(self.automated_tags_dict[package_name]))
+            for tag in tags_missing:
+                dist = 1
+                for tag_name, confidence_score in self.sorted_automated_tags[package_name]:
+                    if tag == tag_name:
+                        total_sum += dist * dist
+                        break
+                    else:
+                        dist += 1
+        return math.sqrt(total_sum)
```

Comment:
I'm not sure with this:

1. I think this could resolve in per-run specific results as there is no guarantee that tags_missing are sorted anyhow - and even if they would be the calculation does not make sense to me (why `break`)?

2. Could you please elaborate on this euclidean distance calculation?  I'm not sure with formula you use.

Reply:
Suppose

`manual tags = ["one", six", "five"]`

`result = ["one", "two", "three", "four"]`

`_sorted = [["one", 0.99], ["two", 0.98], ["three", 0.97], ["four", 0.96], ["six", 0.6], ["five",0.5]]`

So we calculate an error measure as follows:

Since there is no ordering to manual tags, for every tag we take it that the tag should come first in the
sorted sequence. By this definition, this error metric can never be zero.

For tag "six", calculating error measure as position from top:

total = (0 - 5) ^ 2

Similar for tag "five":

total += (0 - 6) ^ 2

euclidean distance = sqrt(total) = sqrt(5*5 + 6*6)

Reply:
I think with this you do not penalize tags that are not present. I thought we would like to compute euclidean distance in orthonormal cube (tag present = 1, tag not present = 0) - weights were there just to do thresholding originally. I'm curious about results of this approach though - do you want to proceed to merge?

Reply:
> weights were there just to do thresholding originally

Right, so my idea behind this metric was to make sure that any work we do around thresholding doesn't actually take us "backwards" as in this score should improve. I wanted to have a metric for the thresholding work. I wasn't targeting packages where we didn't retrieve all the manual tags at all here.

> I thought we would like to compute euclidean distance in orthonormal cube (tag present = 1, tag not present = 0)

We can do that as well, but I think that metric would not apply in the case I just mentioned, since it's for packages where all tags are present in sorted.

Reply:
I think I can do the orthogonal distance metric in a separate PR. This one can be merged.